File: lib/outcome-utils.ts
/**
 * 推薦結果追跡（RecommendationOutcome）ユーティリティ
 *
 * 推薦1件ごとに、推薦時点の状態と時間経過後のリターンを記録する。
 * 精度検証用のデータ基盤。
 */

import { prisma } from "@/lib/prisma"
import { getSectorTrend } from "@/lib/sector-trend"
import { getSectorGroup } from "@/lib/constants"

export type OutcomeType = "daily" | "purchase" | "analysis"
export type Prediction = "buy" | "stay" | "remove" | "up" | "down" | "neutral"

export interface OutcomeData {
  type: OutcomeType
  recommendationId: string
  stockId: string
  tickerCode: string
  sector?: string | null
  recommendedAt: Date
  priceAtRec: number
  prediction: Prediction
  confidence?: number | null
  volatility?: number | null
  marketCap?: bigint | null
  sectorTrendScore?: number | null
  sectorTrendDirection?: string | null
}

/**
 * RecommendationOutcome レコードを作成する。
 * 失敗しても推薦自体の保存には影響させない（エラーはログに出力）。
 */
export async function insertRecommendationOutcome(
  data: OutcomeData
): Promise<string | null> {
  try {
    // セクタートレンドの自動取得（呼び出し元が未指定の場合）
    let trendScore = data.sectorTrendScore ?? null
    let trendDirection = data.sectorTrendDirection ?? null
    const sectorGroup = data.sector ? getSectorGroup(data.sector) ?? data.sector : null
    if (trendScore === null && sectorGroup) {
      try {
        const trend = await getSectorTrend(sectorGroup)
        if (trend) {
          trendScore = trend.compositeScore
          trendDirection = trend.trendDirection
        }
      } catch {
        // トレンド取得失敗はoutcome作成をブロックしない
      }
    }

    const outcome = await prisma.recommendationOutcome.upsert({
      where: {
        type_recommendationId: {
          type: data.type,
          recommendationId: data.recommendationId,
        },
      },
      create: {
        type: data.type,
        recommendationId: data.recommendationId,
        stockId: data.stockId,
        tickerCode: data.tickerCode,
        sector: data.sector,
        recommendedAt: data.recommendedAt,
        priceAtRec: data.priceAtRec,
        prediction: data.prediction,
        confidence: data.confidence,
        volatility: data.volatility,
        marketCap: data.marketCap,
        sectorTrendScore: trendScore,
        sectorTrendDirection: trendDirection,
      },
      update: {
        stockId: data.stockId,
        tickerCode: data.tickerCode,
        sector: data.sector,
        recommendedAt: data.recommendedAt,
        priceAtRec: data.priceAtRec,
        prediction: data.prediction,
        confidence: data.confidence,
        volatility: data.volatility,
        marketCap: data.marketCap,
        sectorTrendScore: trendScore,
        sectorTrendDirection: trendDirection,
      },
    })

    return outcome.id
  } catch (error) {
    console.error("Failed to insert RecommendationOutcome:", error)
    return null
  }
}

File: scripts/lib/outcome_utils.py
"""
推薦結果追跡（RecommendationOutcome）ユーティリティ

推薦1件ごとに、推薦時点の状態と時間経過後のリターンを記録する。
精度検証用のデータ基盤。
"""

from datetime import datetime
from decimal import Decimal
from typing import Any


def insert_recommendation_outcome(
    conn,
    outcome_data: dict[str, Any]
) -> str | None:
    """
    RecommendationOutcome レコードを作成する。

    Args:
        conn: psycopg2接続オブジェクト
        outcome_data: 以下のキーを持つ辞書
            - type: "daily" | "purchase" | "analysis"
            - recommendationId: 元の推薦レコードのID
            - stockId: 銘柄ID
            - tickerCode: ティッカーコード
            - sector: セクター（nullable）
            - recommendedAt: 推薦日時
            - priceAtRec: 推薦時点の株価
            - prediction: "buy" | "stay" | "remove" | "up" | "down" | "neutral"
            - confidence: 信頼度 0.0-1.0（nullable）
            - volatility: ボラティリティ（nullable）
            - marketCap: 時価総額（nullable）

    Returns:
        作成されたレコードのID。失敗時はNone。
    """
    try:
        with conn.cursor() as cur:
            cur.execute(
                '''
                INSERT INTO "RecommendationOutcome" (
                    id,
                    type,
                    "recommendationId",
                    "stockId",
                    "tickerCode",
                    sector,
                    "recommendedAt",
                    "priceAtRec",
                    prediction,
                    confidence,
                    volatility,
                    "marketCap",
                    "createdAt",
                    "updatedAt"
                ) VALUES (
                    gen_random_uuid(),
                    %s, %s, %s, %s, %s,
                    %s, %s, %s, %s, %s, %s,
                    NOW(), NOW()
                )
                ON CONFLICT (type, "recommendationId") DO NOTHING
                RETURNING id
                ''',
                (
                    outcome_data["type"],
                    outcome_data["recommendationId"],
                    outcome_data["stockId"],
                    outcome_data["tickerCode"],
                    outcome_data.get("sector"),
                    outcome_data["recommendedAt"],
                    outcome_data["priceAtRec"],
                    outcome_data["prediction"],
                    outcome_data.get("confidence"),
                    outcome_data.get("volatility"),
                    outcome_data.get("marketCap"),
                )
            )
            result = cur.fetchone()
            conn.commit()

            if result:
                return result[0]
            return None

    except Exception as e:
        print(f"  Warning: Failed to insert RecommendationOutcome: {e}")
        conn.rollback()
        return None
